django/db/models/sql/compiler.py

```diff
@@ -387,18 +387,19 @@ def _order_by_pairs(self):
                     True,
                 )
                 continue
-            if col in self.query.annotations:
-                # References to an expression which is masked out of the SELECT
-                # clause.
-                if self.query.combinator and self.select:
+
+            ref, *transforms = col.split(LOOKUP_SEP)
+            if expr := self.query.annotations.get(ref):
+                if transforms:
+                    for name in transforms:
+                        expr = self.query.try_transform(expr, name)
+                elif self.query.combinator and self.select:
                     # Don't use the resolved annotation because other
                     # combinated queries might define it differently.
-                    expr = F(col)
-                else:
-                    expr = self.query.annotations[col]
-                    if isinstance(expr, Value):
-                        # output_field must be resolved for constants.
-                        expr = Cast(expr, expr.output_field)
+                    expr = F(ref)
+                if isinstance(expr, Value):
+                    # output_field must be resolved for constants.
+                    expr = Cast(expr, expr.output_field)
```

Comment:
A transform could theoretically generate a `Value` expression even if unlikely so I figured I'd add it here.